### DISCUSSION TITLE
Handle URL's / protocols

### DISCUSSION THREAD

**Discussion by tml** [ghostty-org/ghostty#3021] [Feature Requests, Ideas]
It would be nice to have some protocol handlers registered in ghostty, so that it can be the default application for certain protocols/URLs. As a concrete example, I would like to be able to have links using the "ssh" protocol (e.g., `ssh://user@host`) be able to:
1. Focus my open Ghostty terminal
2. Create a new window / tab / split
3. Begin an ssh session using the `user@host` portion of the URL

A few things that need to be considered:

- How do we safely handle the URL? iTerm2 [has some interesting thoughts](https://github.com/gnachman/iTerm2/blob/885a8a89146262de7f05a21ddf9bba935c35c7cb/sources/NSURL%2BiTerm.m#L157-L180) about handling it that we might want to learn a lesson or two from
- How do I tell ghostty what I want it to do when it gets a URL (what helper applications should it leverage? How does it parse the URL fragments? Does it open a new window, a new tab, or split the current view?)
- Which protocols should we handle in ghostty? Are there reasonable default assumptions we can make for those protocols to the above question?

Here's a screenshot of the URL Scheme handling options of current iTerm2
![image](https://github.com/user-attachments/assets/41f8cfbd-d3f4-41bd-b9bd-be3a24765fc0)

**Comment by westerlind** [comment]
This feature is available in both iTerm2 and Terminal.app (at least for ssh:// ), so it feels like something that is expected to be available on a native macOS terminal.
On Linux, too few terminals have this feature, so I have been missing it enough to write my own wrapper script to get this functionality there, for whatever terminal I have been using at the moment.
That should be quite easy to do for Ghostty on Linux as a workaround (at least if opening a new Ghostty window rather than a new tab is good enough), but not on macOS as it's not currently possible to launch Ghostty and make it run a specified command there.

**Comment by PhilETaylor** [comment]
<del>This is the reason Im having to go back to iTerm today :-( 

<del>I use Alfred with a custom script that allows me to launch ssh sessions to one of over 30 servers (cmd space, ssh, servername, enter and then this launches the terminal with the correct ssh command and flags) 

<del>With no way to set ghostty as the default handler for ssh:// urls, no way to make Alfred use ghostty as the terminal, and no way to pass ghostty cli a url to open, it means I cannot launch ghostty and init a ssh session to a remote server. 

<del>When you login to multiple servers multiple times a day this kind of automation is king. :( 

EDIT:

It seems the following script put in Alfred -> Features -> Terminal works :) 🎉🎉🎉
Thanks to ChatGPT and the examples in https://github.com/vitorgalvao/custom-alfred-iterm-scripts/blob/master/custom_iterm_script.applescript for the hints. 

```
on alfred_script(query)
  do shell script "open -a Ghostty"
  tell application "Ghostty" to activate
  tell application "System Events"
	keystroke query
    keystroke return
  end tell
  return
  
end alfred_script
```

**Comment by rhodes-b** [comment]
https://github.com/ghostty-org/ghostty/discussions/2394

this should solve your problem by allowing you to handle ssh commands however you want, opening in a new tab isn't directly supported but starting a new window definitely should be (if you just tell it to run ghostty)

controlling a new tab would be included in #2353
